### DISCUSSION TITLE
Do you plan to rewrite the GUI with MAUI to be able to run on Linux/MacOs

### DISCUSSION THREAD

**Discussion by vricosti** [gitextensions/gitextensions#10305] [General]
Hi,

GitExtension is the best tool for git and I am sad when using linux or macos, any plan to make it work on these platforms ?
By the way I don't understand why in 2022 developers still bind logic and UI so tightly. 
It's better to make a kind of backend that discuss with the UI through either callbacks or websocket, that's how we have transformed an old desktop application by splitting the logic and we rewrote the ui in web.

**Comment by gerhardol** [comment]
There are no active efforts that I know about rewriting the application. Other alternatives mentioned are Avalonia.
Using wine and a file path translation handling similar to WSL is probably a simpler way forward. https://github.com/gitextensions/gitextensions/issues/6051#issuecomment-1143459558
As the project is a volunteer project any of these changes should be steered by someone that has the interest to see the features. None in the core team needs such features what I know. To rewrite the application would be a major work, probably several man years.
The lack of active Mono contributors together with the stop of development of Mono resulted in removed support for Mono.

**Comment by nedy13** [comment]
Can GitExtension version 4.x run on Linux again, because .Net6 runs on Linux, too (other than .Net 4.x.) Is there any limitation which prevents this?

**Reply by gerhardol** [reply]
WinForms requires Windows, so the answer is unfortunately no.

**Reply by RussKie** [reply]
.NET  is a cross-platform SDK, but the app is built on Windows Desktop .NET, which is Windows-only.

Windows Forms in .NET 7 has enhanced databinding support, and, in theory, it'll make it easier to restructure the app and embrace MVVM pattern. This, in turn, will allow writing a new UI layer, if desired.

This is not on the team's roadmap,  so contributions are welcome.

**Reply by telyakov** [reply]
> .NET  is a cross-platform SDK, but the app is built on Windows Desktop .NET, which is Windows-only.
> 
> 
> 
> Windows Forms in .NET 7 has enhanced databinding support, and, in theory, it'll make it easier to restructure the app and embrace MVVM pattern. This, in turn, will allow writing a new UI layer, if desired.
> 
> 
> 
> This is not on the team's roadmap,  so contributions are welcome. 

Team must prepare solution for multi platform ui, maybe extract ui to declarative independent approach.

**Reply by vbjay** [reply]
> > .NET  is a cross-platform SDK, but the app is built on Windows Desktop .NET, which is Windows-only.
> > 
> > 
> > 
> > Windows Forms in .NET 7 has enhanced databinding support, and, in theory, it'll make it easier to restructure the app and embrace MVVM pattern. This, in turn, will allow writing a new UI layer, if desired.
> > 
> > 
> > 
> > This is not on the team's roadmap,  so contributions are welcome. 
> 
> Team must prepare solution for multi platform ui, maybe extract ui to declarative independent approach.

@telyakov 

This is a story about four people named Everybody, Somebody, Anybody and Nobody. There was an important job to be done and Everybody was sure that Somebody would do it. Anybody could have done it, but Nobody did it. Somebody got angry about that, because it was Everybody’s job. Everybody thought Anybody could do it, but Nobody realized that Everybody wouldn’t do it. It ended up that Everybody blamed Somebody when Nobody did what Anybody could have.

@RussKie and other core team members are volenteers. I and other contributors push changes to make things better and solve issues. Your must is not needed. Yes we have a desire to make it xplat but it is a huge push. We have had a few conversations about who h platform and Maui is one. Yes the project is separated out so gui uses the commands library. But rewriting the gui into some non winforms is a HUGE lift. So if you want to contribute and give assistance on making that lift.....by all means experiment, try creating a maui copy of the gui project. Put some effort into it instead of being an anybody be a somebody. 

Could start with a create tag sample project to see if you can create tags in a maui sample using the command library.

**Reply by mbostwick** [reply]
@vbjay the problem and concern is the core team hasn't given a direction. In the past I tried to request[ in issue form](https://github.com/gitextensions/gitextensions/issues/8422) updating a small UI feature on commits, and got completely shut down because the core team didn't think it was a feature they wanted (I was willing to do the work at the time). 

There are lots of projects and peoples time are always limited. So trying to start the conversation and work things out is reasonable.. 

Honestly, this project has been on my mind alot, because git has become such an important thing in so many peoples workflows. 

@telyakov you might consider writing a UI for https://github.com/Byron/gitoxide that would allow it to support web assembly in the long wrong, and the UI could run in a browser or run locally, which would be very powerful.

This would also allow this team to target blazer and use the same UI. (I should also add, that there are other people who have started building web UI's for git, so integrating into those would reduce the lift, and be reasonably possible)

**Reply by vbjay** [reply]
@mbostwick 
Sounds cool. Did you actually create a pull request and show what you meant? Sometimes creating the issue and pull request helps expres how it would work.  Some features are not desired. Some are covered by other features. 

Yes the convo is fine. The core team and contributors want the project to be better. But we get a lot of this project should without a lot of somebody putting fingers to keys and actually lifting. So I think if someone felt Maui should be the way to go try writing a sample using one of the easy forms like create tag to see what that would look like. A full rewrite of the whole ui would take a long time so we have had some suggestions on different ui frameworks. So a lot of it is concern of support. Maui is more likely because it is part of .net. But we have seen wpf, uwp and such move a lot and so I think.some of it is winforms is very stable while Maui still seems like a lot of chrome and not a lot of steel behind it. We don't want to rework into maui and be stuck with a lack of reliability or losing wide os suport. We got bit by this with mono. So it's not that we don't want Maui. It's that we want to tread carefully and not lose stability.

**Reply by mbostwick** [reply]
I did not create a PR, because I have found that getting consensus for a feature, technology, and an implementation should come before spending time on the work.  Otherwise, you can end up spending time on something that returns no value.

**Reply by gerhardol** [reply]
At least me do not have the deep knowledge about other GUI to determine the benefits of the alternatives for a multiplatform GUI.
This would be cool, but not something I am in need of myself and I would not be the driver of such a feature.

For a project like GE, the primary motivation for contributors and maintainers are to get cool features, that is what you spend your happy time on.
More manhours are needed to maintain GE, the 4.0 release took quite some time. The normal support takes quite some time. 
So personally, even if I am interested in multiplatform, I do not know if I can contribute much to the direct conversion.

New maintainers are welcome.
There will not be a consensus before a work is started, but some (rough) plans is a good start to start a conversion.

**Reply by RussKie** [reply]
> Team must prepare solution for multi platform ui, maybe extract ui to declarative independent approach.

@telyakov I not finding the tone of your post friendly or polite. The team doesn't "must" or "have to" or "owe" anything to anyone. The team consists of people who selflessly donate their time and energy to build and maintain the app and all other projects associated with it.

> the problem and concern is the core team hasn't given a direction.

@mbostwick to give a direction there has to be a technical discussion. However, instead of technical discussions with proposals exploring options, weighing pros and cons, etc. users do a drive-by requests or demands that the team must do something (see above). For example, in .NET 7 Windows Forms has [enhanced data-bindings](https://devblogs.microsoft.com/dotnet/winforms-enhancements-in-dotnet-7/#databinding-improvements) that allow moving Windows Forms apps to the MVVM pattern and, thus, making possible to restructure the code to support another UI stack in the future. 


> In the past I tried to request[ in issue form](https://github.com/gitextensions/gitextensions/issues/8422) updating a small UI feature on commits, and got completely shut down because the core team didn't think it was a feature they wanted (I was willing to do the work at the time).

Like with any software - big or small - there are constraints (such as people, time, budget, etc). And every feature goes through a prism of value-add. Making changes to rarely used features could often lead to negative value-add because it would require additional maintenance.
My _personal_ view is that that specific dialog is largely irrelevant - e.g., how many developers create git repos from scratch without copying established patterns from somewhere else? I haven't seen a single one in the past decade.

**Reply by mbostwick** [reply]
If I'm honest, it sounds like the consensus is that most people would accept PR's that move things to MVVM. However, since the core team is all windows, theres no driving force to make the tool multi-platform, so the work would never really get merged, which is completely okay.

For me personally I've been looking for a good tool to suggest in helping people get better at git, which git extensions does great for windows, and I now work at a place where all the devs use apple (this seems to be a trend at a lot of places, from my perspective..) , so I haven't really found a good answer for jr devs.  Kudo's of course for the hard work, and git extensions is a great tool. 

Honestly I would love to see the team support 2 git projects rather then 1, with 1 project being an nuget project for dotnet git client, and the other project being the UI project. As I said though, I don't want to waste time doing something unless I've got buy in from people that its time well spent, and I personally find value in it. I assume other devs feel the same. A good community makes a project that much better. Honestly the core team on this project seems engaged and does alot of work to get things tagged. 

I think what your calling drive by requests is more of a work flow, where someone goes man, it would be great if this tool could do x, then route a feature request. With positive feedback a person may just go ahead and do it, or maybe they will point someone towards the issue and say if you want to learn more about git, why not try working on that feature. I don't think being cranky with people for routing requests and having discussions is a great use of time, if you don't want to do something,  you can simply ignore it, or suggest an alternative path.

**Reply by RussKie** [reply]
> However, since the core team is all windows, theres no driving force to make the tool multi-platform, so the work would never really get merged, which is completely okay.

I wouldn't put it so categorically. Changes that benefit the product are considered and, generally, are accepted.

> As I said though, I don't want to waste time doing something unless I've got buy in from people that its time well spent, and I personally find value in it.

I think that goes for everyone. 
I also hate seeing contributors spent their time and efforts making changes without talking to the team and then seeing the changes rejected (this is not specific to this project).

> I think what your calling drive by requests is more of a work flow, where someone goes man, it would be great if this tool could do x, then route a feature request. With positive feedback a person may just go ahead and do it

This is, sadly, not my general experience - neither here nor in other repos I'm involved. There are a lot of people out of there that take software for granted and have entitlement attitude. With that they open drive-by FR and then some days/weeks/months later come and ask "Has it been done yet?".

If you really feel passionate about a specific feature - champion it! Open a new or reuse an existing thread to discuss goals and possible ways to get there, create a POC and take everyone else on that journey.

**Comment by mbostwick** [comment]
What does the core team think of Avalonia ?

**Reply by gerhardol** [reply]
I have no opinion, with limited experience in any of the possible toolkits.
